### PR TITLE
Use the absolute path when using odo push

### PR DESCRIPTION
**Pull request by cdrage**
This PR fixes the following issues:
  - Deletion not working within windows and using a non-relative path
  command, ex: `odo push --context nodejs-ex/`
  - File changes not working when using a non-relative path command:
  `odo push --context nodejs-ex/`
  - Use forward slashes when storing information into
  odo-file-index.json

To test (on Windows):
```sh
git clone https://github.com/openshift/nodejs-ex
cd nodejs-ex\
odo create nodejs
odo push
cd ..
odo push --context nodejs-ex
type nul > nodejs-ex\foobar.txt
odo push --context nodejs-ex
```

All of these commands should now work as odo will now pass on the
absolute path to the file index functions and commands before deletion /
propagating changed files

Then check the index:
```sh
type nodejs-ex\.odo\odo-file-index.json
```

Which will now have forward slashes.

Closes issue https://github.com/openshift/odo/issues/2301
Closes issue https://github.com/openshift/odo/issues/2217

### CODE REVIEW

**Comment by cdrage** [comment]
This was a pain in the butt to diagnose, but may as well push it and see if it passes everything....

/hold

TODO: Write tests

**Comment by cdrage** [comment]
/hold cancel

Ready for review! Wrote some tests @amitkrout

**Comment by cdrage** [comment]
Another flake:

```sh
Lease acquired, installing...
Installing from release registry.svc.ci.openshift.org/ci-op-9nx58imz/release@sha256:743a170fdd5f3ccc75b2f7218f945167eb910f0b789f10fa311d1382c33c6e01
level=fatal msg="failed to fetch Terraform Variables: failed to fetch dependency of \"Terraform Variables\": failed to fetch dependency of \"Bootstrap Ignition Config\": failed to fetch dependency of \"Common Manifests\": failed to generate asset \"DNS Config\": getting public zone for \"origin-ci-int-aws.dev.rhcloud.com\": listing hosted zones: Throttling: Rate exceeded\n\tstatus code: 400, request id: c7209482-e777-4a54-8124-82a42acb86b4"
```

/test v4.2-e2e-scenarios

**Comment by girishramnani** [comment]
/approve 
Would be great if we had some unit tests for the functions that were changed. Rest all works

**Comment by cdrage** [comment]
@girishramnani So until https://github.com/openshift/odo/pull/2311 is merged, I'll be unable to write filesystem tests (since we will have to "fake" the filesystem to test this out with regards to unit tests).

For now I think just having an e2e test would be okay since this is very context / filesystem specific.

**Comment by cdrage** [comment]
Still need more reviews from someone other than @girishramnani 

Ping @dharmit @amitkrout @mohammedzee1000 @kanchwala-yusuf @kadel

**Comment by cdrage** [comment]
Ping @dharmit @amitkrout @mohammedzee1000 @kanchwala-yusuf @kadel

**Comment by cdrage** [comment]
Ping @dharmit @amitkrout @mohammedzee1000 @kanchwala-yusuf @kadel

File: pkg/odo/cli/component/common_push.go [hunk 1]
```diff
@@ -29,10 +29,10 @@ type CommonPushOptions struct {
...
 	localConfigInfo  *config.LocalConfigInfo
 
-	pushConfig  bool
-	pushSource  bool
-	forceBuild  bool
-	isCmpExists bool
+	pushConfig         bool
+	pushSource         bool
+	forceBuild         bool
+	doesComponentExist bool
```
Comment (mohammedzee1000) [inline pkg/odo/cli/component/common_push.go]: unrelated change? I guess it makes sense in itself, but it is not related to PR topic

File: tests/integration/cmd_push_test.go [hunk 2]
```diff
@@ -56,6 +59,34 @@ var _ = Describe("odo push command tests", func() {
...
+	})
+
+	Context("Test push outside of the current working direcory", func() {
+
+		It("Push, modify a file and then push outside of the working directory", func() {
+			helper.CmdShouldPass("git", "clone", "https://github.com/openshift/nodejs-ex", context+"/nodejs-ex")
+			helper.CmdShouldPass("odo", "component", "create", "nodejs", cmpName, "--project", project, "--context", context+"/nodejs-ex", "--app", appName)
+
+			// Change to the context directory to make sure we push *outside* of the current directory
+			helper.Chdir(context)
```
[the remark this answers is not in the document]
Comment (cdrage) [reply tests/integration/cmd_push_test.go]: Updated@

**Comment by cdrage** [comment]
Updated the tests @kadel !

File: pkg/odo/cli/component/common_push.go (hunk 1 above)
Comment (cdrage) [reply pkg/odo/cli/component/common_push.go]: Yeah, I didn't like the vague variable name before haha.

File: tests/integration/cmd_push_test.go [hunk 3]
```diff
@@ -56,6 +59,42 @@ var _ = Describe("odo push command tests", func() {
...
+			currentWorkingDirectory = helper.Getwd()
+			helper.Chdir(context)
+		})
+
+		// Change back to the curentWorkingDirectory
+		var _ = AfterEach(func() {
+			helper.Chdir(currentWorkingDirectory)
+		})
+
+		FIt("Push, modify a file and then push outside of the working directory", func() {
```
Comment (amitkrout) [inline tests/integration/cmd_push_test.go]: FIt -> It

Comment (cdrage) [reply tests/integration/cmd_push_test.go]: I ALWAYS miss this. Woops.

File: tests/integration/cmd_push_test.go [hunk 4]
```diff
@@ -56,6 +59,42 @@ var _ = Describe("odo push command tests", func() {
...
+
+	Context("Test push outside of the current working direcory", func() {
+
+		// Change to "outside" the directory before running the below tests
+		var _ = BeforeEach(func() {
+			currentWorkingDirectory = helper.Getwd()
+			helper.Chdir(context)
+		})
+
+		// Change back to the curentWorkingDirectory
```
Comment (mik-dass) [inline tests/integration/cmd_push_test.go]: `currentWorkingDirectory`

**Comment by cdrage** [comment]
@mik-dass Updated!

**Comment by mik-dass** [comment]
```
[odo] ---> Installing application source
[odo] + echo '---> Installing application source'
[odo] + cp -Rf /tmp/src/. ./
[odo] ---> package.json modified
[odo] + '[' true == true ']'
[odo] + '[' -f /tmp/refresh//package.json.sha1 ']'
[odo] + echo '---> package.json modified'
[odo] + sha1sum package.json
[odo] sha1sum: package.json: No such file or directory
```
@cdrage the scenario `Push, modify a file and then push outside of the working directory` is failing with the above build log

**Comment by cdrage** [comment]
> ```
> [odo] ---> Installing application source
> [odo] + echo '---> Installing application source'
> [odo] + cp -Rf /tmp/src/. ./
> [odo] ---> package.json modified
> [odo] + '[' true == true ']'
> [odo] + '[' -f /tmp/refresh//package.json.sha1 ']'
> [odo] + echo '---> package.json modified'
> [odo] + sha1sum package.json
> [odo] sha1sum: package.json: No such file or directory
> ```
> 
> @cdrage the scenario `Push, modify a file and then push outside of the working directory` is failing with the above build log

Flake? I'm really not sure.

/retest

**Comment by mik-dass** [comment]
```
Running odo with args [odo project create fplfufgato -w -v4]
[odo] I1206 17:40:05.138295   24134 cmdutils.go:48] Error:
[odo] unable to get odo config file: getwd: no such file or directory
[odo]  ✗  getwd: no such file or directory
Deleting project: qxbjtscfxg
```
/retest

**Comment by cdrage** [comment]
Terraform errors..

/retest
